### ISSUE TITLE
No H264 video in desktop mode with Gstreamer

### ISSUE THREAD

**Issue by TijZwa** [Xpra-org/xpra#3744]
**Describe the bug**
I do not receive H.264 frames with monitors mode, however Gstreamer is enabled.

**To Reproduce**
1. `XPRA_GSTREAMER_ENCODER_PLUGIN=vaapih264enc xpra monitor :10 --bind-ws=0.0.0.0:10000 --bind-tcp=0.0.0.0:10001 --video-encoders=gstreamer --resize-display=1600x1200 -d video,score`
2. Connect with the Xpra 5.x client on windows.
3. Never receive a H.264 frame

**System Information (please complete the following information):**
 - Server OS: Fedora 37
 - Client OS:  Windows 11
 - Xpra Server Version 5.0-10.r32480.fc37
 - Xpra Client Version 5.0 revision 32459

**Log**
Gstreamer seems to be unknown:
```
2023-01-19 16:19:49,313 set_modules(['gstreamer'], ['csc_swscale', 'csc_libyuv'], ())
2023-01-19 16:19:49,313 Warning: ignoring unknown video encoders: gstreamer
2023-01-19 16:19:49,313 VideoHelper.set_modules('gstreamer', 'csc_swscale, csc_libyuv', '') video encoders=gstreamer, csc=csc_swscale, csc_libyuv, video decoders=
```

But later:
```
2023-01-19 16:19:50,111 VideoHelper.init()
2023-01-19 16:19:50,111 VideoHelper.init() initialized=False
2023-01-19 16:19:50,111 init_video_encoders_options()
2023-01-19 16:19:50,111  will try video encoders: gstreamer
2023-01-19 16:19:50,154 xpra is ready.
2023-01-19 16:19:50,226 unhandled bus message type <flags GST_MESSAGE_HAVE_CONTEXT of type Gst.MessageType>: <Gst.Message object at 0x7fabc77abb90 (GstMessage at 0x7fabd84474c0)>
2023-01-19 16:19:50,307  encoder for gstreamer: enc_gstreamer
2023-01-19 16:19:50,307 init_video_encoder_option(enc_gstreamer)
2023-01-19 16:19:50,307  module=<module 'xpra.codecs.gstreamer.encoder' from '/usr/lib64/python3.11/site-packages/xpra/codecs/gstreamer/encoder.py'>
2023-01-19 16:19:50,307     gstreamer encodings=h264, av1
2023-01-19 16:19:50,307  gstreamer  input colorspaces for  h264: YUV420P
2023-01-19 16:19:50,307  gstreamer  input colorspaces for   av1: YUV420P
2023-01-19 16:19:50,307 video encoder options: {'h264': {'YUV420P': [gstreamer(YUV420P to h264)]}, 'av1': {'YUV420P': [gstreamer(YUV420P to av1)]}}
2023-01-19 16:19:50,307 found 2 video encoders: h264, av1
2023-01-19 16:19:50,307 init_csc_options()
2023-01-19 16:19:50,307  will try csc modules: csc_swscale, csc_libyuv
2023-01-19 16:19:50,324 init_csc_option(csc_swscale)
2023-01-19 16:19:50,324  module=<module 'xpra.codecs.ffmpeg.colorspace_converter' from '/usr/lib64/python3.11/site-packages/xpra/codecs/ffmpeg/colorspace_converter.cpython-311-x86_64-linux-gnu.so'>
```

**Comment by ehfd** [comment]
You too?

**Comment by ehfd** [comment]
https://github.com/selkies-project/selkies-gstreamer/issues/78

**Comment by TijZwa** [comment]
> You too?

@ehfd but only with `start monitor`, not in seamless mode.

**Comment by ehfd** [comment]
Ah, not in web browsers.

**Comment by TijZwa** [comment]
> Ah, not in web browsers.

Well, I hit the issue with the GTK/Python client and my own C# implementation. Pretty sure it's a server side issue and will also occur in browsers (xpra-html5)

**Comment by TijZwa** [comment]
```
2023-01-23 13:38:51,502 get_video_pipeline_options(('h264',), 1920, 1200, 'BGRX') scores=[(3, (1, 1), (1, 1), 1920, 1200, libyuv(BGRX to YUV420P), 'YUV420P', (1, 1), 1920, 1200, gstreamer(YUV420P to h264)), (-9, (1, 1), (1, 1), 1920, 1200, swscale(BGRX to YUV420P), 'YUV420P', (1, 1), 1920, 1200, gstreamer(YUV420P to h264))]
2023-01-23 13:38:51,502 check_pipeline_score(False) best=(3, (1, 1), (1, 1), 1920, 1200, libyuv(BGRX to YUV420P), 'YUV420P', (1, 1), 1920, 1200, gstreamer(YUV420P to h264))
2023-01-23 13:38:56,711 get_video_pipeline_options(('h264',), 1920, 1200, 'BGRX', False) last params=(('h264',), 1920, 1200, 'BGRX'), full_csc_modes=typedict({'webp': ('BGRX', 'BGRA'), 'jpeg': ('BGRX', 'BGRA', 'BGR', 'RGBX', 'RGBA', 'RGB', 'YUV420P', 'YUV422P', 'YUV444P'), 'h264': ('YUV420P',), 'vp8': ('YUV420P',), 'vp9': ('YUV420P',)})
2023-01-23 13:38:56,712 get_video_pipeline_options(('h264',), 1920, 1200, 'BGRX') speed: 6 (min 1), quality: 99 (min 1)
2023-01-23 13:38:56,712  no matching colorspace specs for BGRX - direct (no csc)
2023-01-23 13:38:56,712  no matching colorspace specs for RGB - via RGB
2023-01-23 13:38:56,712  no matching colorspace specs for BGR - via BGR
2023-01-23 13:38:56,712  no matching colorspace specs for XRGB - via XRGB
2023-01-23 13:38:56,712  no matching colorspace specs for RGBX - via RGBX
2023-01-23 13:38:56,713  no matching colorspace specs for XBGR - via XBGR
```

**Comment by TijZwa** [comment]
I found that video dection bails out with `not enough frames (0)` in monitor mode.
See https://github.com/Xpra-org/xpra/blob/d85c0475cc3ccd3ef9d0b8c11e5ba244c34c97ad/xpra/server/window/window_video_source.py#L517

If I change https://github.com/Xpra-org/xpra/blob/d85c0475cc3ccd3ef9d0b8c11e5ba244c34c97ad/xpra/server/window/window_video_source.py#L478 line to `video_hint = True`, the H264 frames start comming in.

**Comment by totaam** [comment]
@TijZwa this will lower the threshold for using a video encoder.
We should also lower it when a hardware accelerated encoder is available.

**Comment by TijZwa** [comment]
@totaam yes, this is a dirty fix to prove the issue is in video detection.
Please see the log below when I watch a youtube video in monitor-mode:

```
[vpo@fedora ~]$ tail -f /run/user/1000/xpra/10/server.log | grep nonvideo
2023-01-25 09:19:20,309 nonvideo(None, not enough frames)
2023-01-25 09:19:20,409 nonvideo(None, not enough frames)
2023-01-25 09:19:20,509 nonvideo(None, not enough frames)
2023-01-25 09:19:20,608 nonvideo(None, not enough frames)
2023-01-25 09:19:20,709 nonvideo(None, not enough frames)
2023-01-25 09:19:20,809 nonvideo(None, not enough frames)
2023-01-25 09:19:20,909 nonvideo(None, not enough frames)
2023-01-25 09:19:21,006 nonvideo(None, not enough frames)
2023-01-25 09:19:21,109 nonvideo(None, not enough frames)
2023-01-25 09:19:21,209 nonvideo(None, not enough frames)
2023-01-25 09:19:21,309 nonvideo(None, not enough frames)
2023-01-25 09:19:21,409 nonvideo(None, not enough frames)
2023-01-25 09:19:21,510 nonvideo(None, not enough frames)
2023-01-25 09:19:21,609 nonvideo(None, not enough frames)
2023-01-25 09:19:21,709 nonvideo(None, not enough frames)
2023-01-25 09:19:21,806 nonvideo(None, not enough frames)
2023-01-25 09:19:21,908 nonvideo(None, not enough frames)
2023-01-25 09:19:22,007 nonvideo(None, not enough frames)
2023-01-25 09:19:22,109 nonvideo(None, not enough frames)
2023-01-25 09:19:22,209 nonvideo(None, not enough frames)
2023-01-25 09:19:22,310 nonvideo(None, not enough frames)
2023-01-25 09:19:22,409 nonvideo(None, not enough frames)
2023-01-25 09:19:22,509 nonvideo(None, not enough frames)
```

**Comment by totaam** [comment]
Should now be better: 7e488ebd7437b82a158a2fad8583a6d47f54d844

**Comment by TijZwa** [comment]
I tried installing the latest beta, but still no video..
Please find the log attached.
[server.log](https://github.com/Xpra-org/xpra/files/10756061/server.log)

**Comment by totaam** [comment]
> `must_encode_full_frame(webp)=False full_frames_only=False, non_video=('webp', 'rgb24', 'rgb32', 'jpeg'), video=('h264', 'vp8', 'vp9')`

That's a good set of encodings to work with.

> nonvideo(None, not enough frames)

e096b6f684f1e752dff4057c12b473095f3acc30 should ensure that the `video_hint` is set to either 1 or 2 when a GPU encoder is available (`gpu_cost > cpu_cost`).
This in turn should make it impossible to hit the codepath that leads to `not enough frames`.
But I had messed it up in 3b5c3ebe9575191b8883a19f85170f8e7ba77457 - which disabled the `vaapi` `GStreamer` pipeline, fixed in 1feab02f3d487434475015a5b4f776a197dff943.
New builds in progress...

---

FYI:
> `Error: unknown modifier 'alt, control'`

For now, you need to figure out which modifier `alt` maps to - this is how we do it on MS Windows:
https://github.com/Xpra-org/xpra/blob/75cfd2b7f1944bb2a3ab35460bcf9babc82f72b9/xpra/platform/win32/keyboard.py#L67-L76
Called from:
https://github.com/Xpra-org/xpra/blob/75cfd2b7f1944bb2a3ab35460bcf9babc82f72b9/xpra/client/ui_client_base.py#L519-L529

**Comment by TijZwa** [comment]
> New builds in progress...

Still not working :( I've included the new [server.log](https://github.com/Xpra-org/xpra/files/10760373/server.log)

`xpra info | grep pipeline` gives me:
```
[vpo@fedora ~]$ xpra info | grep pipeline
client.sound.speaker.pipeline=pulsesrc device="auto_null.monitor" name="src" ! timestamp name=timestamp ! queue name=queue min-threshold-time=0 max-size-buffers=0 max-size-bytes=0 max-size-time=50000000000000 leaky=2 ! audioconvert ! removesilence ! audioconvert ! audioresample ! volume name=volume volume=1.0 ! opusenc bitrate-type=1 complexity=0 ! appsink name=sink emit-signals=true max-buffers=10 drop=true sync=false async=false qos=false
client.window.1.encoding.pipeline_last_check=20713
client.window.2.encoding.pipeline_last_check=20714
logging.filters.Encoding.score=Video pipeline scoring and selection
```

> For now, you need to figure out which modifier `alt` maps to - this is how we do it on MS Windows:

Thanks! That's an issue I've to fix in the C# Avalonia client.

**Comment by TijZwa** [comment]
Btw: hardcoding `video_hint` to 2 doesnot work.
I think? the server wants to send video but falls back to webp:

```
2023-02-16 21:31:35,083 make_data_packet: image=XShmImageWrapper(BGRX: 0, 0, 1920, 1080), damage data: (1, 0, 0, 1920, 1080, 'auto')
2023-02-16 21:31:35,083 do_video_encode(auto, XShmImageWrapper(BGRX: 0, 0, 1920, 1080), {'quality': 98, 'speed': 10, 'rgb_formats': ('RGBX', 'RGBA', 'RGB'), 'lz4': True, 'content-type': 'desktop', 'cuda-device-context': None})
2023-02-16 21:31:35,083 fallback encodings(('webp', 'rgb24', 'rgb32', 'jpeg'), {'rgb24': [<function encode at 0x7f508849e980>], 'rgb32': [<function encode at 0x7f508849e980>], 'png': [<cyfunction encode at 0x7f505e724450>, <function encode at 0x7f50805a0e00>], 'png/L': [<cyfunction encode at 0x7f505e724450>, <function encode at 0x7f50805a0e00>], 'png/P': [<function encode at 0x7f50805a0e00>], 'jpeg': [<cyfunction encode at 0x7f5091161a40>, <function encode at 0x7f50805a0e00>], 'webp': [<cyfunction encode at 0x7f5091163030>, <function encode at 0x7f50805a0e00>], 'jpega': [<cyfunction encode at 0x7f5091161a40>]})=('rgb24', 'rgb32', 'jpeg', 'webp')
2023-02-16 21:31:35,120 send_delayed for wid 1, delaying again because too many pixels are waiting to be encoded: 2073600
2023-02-16 21:31:35,131 send_delayed for wid 1, delaying again because too many pixels are waiting to be encoded: 2073600
2023-02-16 21:31:35,141 send_delayed for wid 1, delaying again because too many pixels are waiting to be encoded: 2073600
2023-02-16 21:31:35,143 compress:  60.4ms for 1920x1080 pixels at    0,0    for wid=1     using      webp with ratio   3.2%  ( 8100KB to   257KB), sequence  1056, client_options={'rgb_format': 'BGRX', 'quality': 98, 'encoder': 'xpra.server.window.window_video_source'}, options={'quality': 98, 'speed': 10, 'rgb_formats': ('RGBX', 'RGBA', 'RGB'), 'lz4': True, 'content-type': 'desktop', 'cuda-device-context': None}
2023-02-16 21:31:35,152 send_delayed for wid 1, batch delay is 40ms, elapsed time is 56ms
```

**Comment by totaam** [comment]
> the server wants to send video but falls back to webp:

TLDR: my bad, fixed.

`grep nonvideo server.log` shows:
* lots of `low pixel count` (less than 4KB, often just a few hundred bytes) - so in this case, not using video encoding is definitely the right thing to do: much faster to compress (factor of 10, sometimes 100), much better compression ratio, usually lossless, updates the screen faster
* fewer `not enough frames` - this should not be possible to hit unless `video_hint=0` and this should not be the case if you have a hardware accelerated encoder... but:

The debug output was slightly ambiguous, so I have made some improvements: ae61b42342ffbd71897411fa30866ec39e2fe58e, 182908cf578bb349316cbb2222d679072ce02026, c71c605730df307a14845e77b00230f21cfb6f7b.
But it looks to me like it initialized 2 `h264` encoders:
```
2023-02-16 21:01:43,868 gstreamer: h264 / YUV420P / ('YUV420P',) passed
2023-02-16 21:01:43,963 gstreamer: h264 / YUV420P / ('YUV420P',) passed
```
And since we only have `x264enc` and `vaapih264enc` then the latter should be available and is definitely GPU accelerated.
The new bug I introduced in fixing https://github.com/Xpra-org/xpra/issues/3758#issuecomment-1433026259: this broke in between the previous fixes and 75cfd2b7f1944bb2a3ab35460bcf9babc82f72b9.
9d0c775e528ae5a20db8885cdd1000d432444266 should fix this - finally?

In my testing with Firefox in a small-ish monitor session, it uses mostly a mix of `webp`, `jpeg` and `h264` (via `nvenc` for me). Perhaps using `h264` more than it should in some cases.. But since this is hardware accelerated, that's acceptable.

At least, running with the latest changes should give us (pardon the pun) a better picture.

---

*If* this still does not work for you:
* what command lines did you use?
* did you try with the python client?
* was the screen updating a lot?
* what application are you running? Did you try something simple like `glxgears`? (it should detect a video region and use a video encoding for that subregion at least)

**Comment by TijZwa** [comment]
> [9d0c775](https://github.com/Xpra-org/xpra/commit/9d0c775e528ae5a20db8885cdd1000d432444266) should fix this - finally?

Fixed 🎉 
Thanks!

**Comment by TijZwa** [comment]
This works, but sometimes chrashes my server.
I've attached the backtrace. Is this a Xpra bug? Or is it my system?

H264 works great in seamless mode.
[backtrace.txt](https://github.com/Xpra-org/xpra/files/10812896/backtrace.txt)

**Comment by totaam** [comment]
@TijZwa can you reproduce with the latest git master?
Can you run the server with `-d video,gstreamer`?
